### PR TITLE
crimson/os/seastore/cache: add lba nodes' logical begin addresses to extent_info_t

### PR DESCRIPTION
**Pull request by xxhdx1985126**
LBAManager rely on lba nodes' logical start addresses to judge whether an lba extent
is still live

Fixes: https://tracker.ceph.com/issues/53612
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/cache.cc [hunk 1]
```diff
@@ -996,7 +996,9 @@ record_t Cache::prepare_record(Transaction &t)
 	i->get_type(),
 	i->is_logical()
 	? i->cast<LogicalCachedExtent>()->get_laddr()
-	: L_ADDR_NULL,
+	: (is_lba_node(i->get_type())
+	  ? i->cast<lba_manager::btree::LBANode>()->get_node_meta().begin
+	  : L_ADDR_NULL),
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.cc:1001]: Looks to me `extent_t::addr` has commented an invariant that it is the "laddr of extent (L_ADDR_NULL for non-logical)". And cleaner does use this `extent_info_t::addr` to look up lba extents to check alive. The fix looks good to me, might need @athanatos to take another look.

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by athanatos** [comment]
This addr doesn't seem to actually be used for lba nodes.  BtreeLBAManager::get_physical_extent_if_live does accept a laddr_t parameter, but it just passes it to Cache::get_extent_by_type which ignores it for physical extents.  It would probably be cleaner to remove the laddr_t param from get_physical_extent_if_live.

Perhaps I'm missing something?

**Comment by cyx1231st** [comment]
@athanatos Looks to me `BtreeLBAManager::get_physical_extent_if_live()` passes the laddr to `LBABtree::get_internal_if_live()` and `LBABtree::get_leaf_if_live()`, which further calls `LBABtree::lower_bound()` to lookup the lba tree by the laddr, which is L_ADDR_NULL.

**Comment by cyx1231st** [comment]
> BtreeLBAManager::get_physical_extent_if_live does accept a laddr_t parameter, but it just passes it to Cache::get_extent_by_type which ignores it for physical extents.

Looks to me the problem is later at https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/transaction_manager.cc#L461-L469 when `Cache::get_extent_by_type()` returns nothing, so that `BtreeLBAManager::get_physical_extent_if_live()` will lookup the lba tree using the laddr L_ADDR_NULL.

**Comment by athanatos** [comment]
Ah, I was looking at an older version on my laptop.  Good catch!

**Review by athanatos — APPROVED** [review]

**Pull request merged**